#11 팬풀 기능 구현

dsxl8ior opened the pull request:
### 구현
#11 
- 장소 인증 기능
- 다중 필터링 팬풀 조회 기능 (페이징)
- 팬풀 수정 기능
- 팬풀 참여 기능
- 팬풀 참여 취소 기능
- (응원팀 따라 최신순으로) 홈 팬풀 조회 기능 
- 회원 프로필 조회 시 호스팅한 팬풀 개수 함께 조회되도록 수정
- 회원이 주최하는 팬풀 조회 기능 (페이징)
- 팀이 포함된 야구 게임 조회하는 기능

### 변경 점
- gradle 에 Query DSL 의존성 추가 -> 동적 필터링을 위해
- common JPAConfiguration 변경

dsxl8ior committed:
refactor: 회원 대표동네 추가하는 기능

dsxl8ior committed:
feat: 두 번째 장소까지 인증되게 기능 확장, 대표 장소 수정하는 기능

dsxl8ior committed:
rename: 이번주 야구 경기 목록을 조회하는 서비스

dsxl8ior committed:
refactor: 게임 진행 상태 enum 추가 및 response 구조 변경

dsxl8ior committed:
feat: 팬풀 조회 API에 페이징 기능

dsxl8ior committed:
Merge branch 'feat/advancing-fanpool' of https://github.com/TwoDotFive/server into feat/advancing-fanpool

dsxl8ior committed:
chore: gradle 정렬

dsxl8ior committed:
Merge branch 'main' of https://github.com/TwoDotFive/server into feat/advancing-fanpool

dsxl8ior committed:
chore: git ignore에 generated 폴더 추가

query dsl 관련 Q타입 엔티티 저장됨

dsxl8ior committed:
chore: QueryDsl 관련 의존성 추가

dsxl8ior committed:
feat: 팬풀 다중 필터링 조회 기능

dsxl8ior committed:
fix: 팬풀 title 필드 추가

dsxl8ior committed:
feat: 팬풀 수정 기능

dsxl8ior committed:
feat: 팬풀 모집 상태 수정 기능

dsxl8ior committed:
fix: response 리턴 타입 수정

Team(Entity) -> TeamView(DTO)

dsxl8ior committed:
feat: 회원 프로필 조회에 호스팅한 팬풀 개수 추가

dsxl8ior committed:
fix: 팬풀 상태 변경 기능 클래스명 실수 ^^;

UpdateFanpoolCommand -> UpdateFanpoolStateCommand

dsxl8ior committed:
feat: 다중조건 팬풀 조회할 때 모집 완료된 팬풀은 제외하는 기능

dsxl8ior committed:
feat: 홈에서 사용하는 최신순 팬풀 조회 기능

dsxl8ior committed:
rename: FanpoolController 메서드명 변경

dsxl8ior committed:
rename: FanpoolInformationView -> FanpoolView

dsxl8ior committed:
feat: 팬풀 참여, 참여 취소하는 기능

dsxl8ior committed:
feat: 회원이 호스팅한 팬풀들 최신순으로 조회

dsxl8ior committed:
refactor: dto단 command 패키지 분리

dsxl8ior committed:
feat: Team 별로 Game start date가 최신순으로 조회하는 기능

dsxl8ior committed:
refactor: RequestMapping 수정, SecurityConfig에 RequestMatcher추가

csct3434 commented on src/main/java/com/example/temp/user/domain/UserAuthenticatedLocationRepository.java:
```diff
@@ -1,18 +1,30 @@
+    }
+
+    @Query("SELECT f FROM fanpool f WHERE f.user_id = :user AND f.representative is TRUE ")
+    UserAuthenticatedLocation findByUserAndRepresentative(User user);
```
이거는 Optional 없어도 괜찮나?

csct3434 commented on src/main/java/com/example/temp/user/domain/UserAuthenticatedLocationRepository.java:
```diff
@@ -1,18 +1,30 @@
 
-    Optional<UserAuthenticatedLocation> findByUser(User user);
+    Optional<List<UserAuthenticatedLocation>> findByUser(User user);
 
-    default UserAuthenticatedLocation findByUserOrElseThrow(User user) {
+    default List<UserAuthenticatedLocation> findByUserOrElseThrow(User user) {
         return findByUser(user).orElseThrow(() ->
                 new CustomException(HttpStatus.BAD_REQUEST, "장소 인증을 먼저 진행해 주세요"));
     }
```
조회 결과가 없어도 size가 0인 List가 들어오기 때문에 Optional.isEmpty()는 항상 false로 나옵니당

dsxl8ior replied:
아!!!!!!

csct3434 commented on src/main/java/com/example/temp/user/service/location/impl/FindUserAuthenticatedLocationServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+
+        return new FindUserAuthenticatedLocationResponse(result);
+    }
+}
```
조회 로직이므로 @Trasnactional(readonly = true)를 붙여주는게 데이터 무결성 보장에 좋다고 생각합니돠

csct3434 commented on src/main/java/com/example/temp/user/service/location/impl/UpdateUserAuthenticatedLocationServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+@Transactional
+public class UpdateUserAuthenticatedLocationServiceImpl implements UpdateUserAuthenticatedLocationService {
+    private final UserRepository userRepository;
+    private final UserAuthenticatedLocationRepository userAuthenticatedLocationRepository;
+
+    @Override
+    public void doService(long userId, final long locationId) {
+        User user = userRepository.findByIdOrElseThrow(userId);
+        List<UserAuthenticatedLocation> target = userAuthenticatedLocationRepository.findByUserOrElseThrow(user);
+
+        target.forEach(o -> {
+            o.updateRepresentative(locationId);
+            userAuthenticatedLocationRepository.save(o);
+        });
+    }
```
JPA 변경 감지 덕분에 save 호출 안해도 괜춘!

csct3434 commented on src/main/java/com/example/temp/baseball/domain/value/GameState.java:
```diff
@@ -0,0 +1,34 @@
+package com.example.temp.baseball.domain.value;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+
+@RequiredArgsConstructor
+@Getter
+public enum GameState {
+    BEFORE_START("경기 전"),
+    IN_PROGRESS("경기중"),
+    TERMINATED("경기종료"),
+    CANCELED("취소"),
+    CANCELED_RAIN_OUT("우천시 취소");
+
+    private final String kor;
+
+    public static GameState determineState(LocalDateTime startAt) {
+        LocalDateTime now = LocalDateTime.now();
+
+        // 평균 야구 경기 시간을 3시간 10분 Duration으로 정의
+        Duration gameDuration = Duration.ofHours(3).plusMinutes(10);
+
+        if (now.isBefore(startAt)) {
+            return GameState.BEFORE_START;
+        }
+        if (now.isAfter(startAt) && now.isBefore(startAt.plus(gameDuration))) {
+            return GameState.IN_PROGRESS;
+        }
+        return GameState.TERMINATED;
+    }
+}
```
Enum 분리한거 굿 👍

csct3434 commented on src/main/java/com/example/temp/common/config/JpaConfig.java:
```diff
@@ -1,10 +1,21 @@
 @EnableJpaAuditing
 public class JpaConfig {
 
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @Bean
+    public JPAQueryFactory jpaQueryFactory(EntityManager entityManager) {
+        return new JPAQueryFactory(entityManager);
+    }
```
EntityManager 필드 주입 받을때는 @PersistenceContext 사용하는거 하나 배워가유

근데 jpaQueryFactory() 메서드에서 필드 주입 받은 EntityManager가 아니라 인자로 전달받은 EntityManager 사용하고 있는듯?

dsxl8ior replied:
헉 그렇네요

csct3434 commented on src/main/java/com/example/temp/fanpool/dto/UpdateFanpoolRequest.java:
```diff
@@ -0,0 +1,20 @@
+package com.example.temp.fanpool.dto;
+
+import com.example.temp.geo.dto.AddressInformationView;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDateTime;
+
+@NoArgsConstructor
+@Getter
+public class UpdateFanpoolRequest {
+    private String title;
+    private long gameId;
+    private LocalDateTime departAt;
+    private AddressInformationView departFrom;
+    private String fanpoolType;
+    private String genderConstraint;
+    private int numberOfPeople;
+    private String memo;
+}
```
나도 이번에 알았는데 요청 DTO도 record로 받을 수 있더라
매번 불필요하게 @NoArgsConstructor 안붙여도 되고 불변성도 보장할 수 있어서 좋은듯

csct3434 commented on src/main/java/com/example/temp/user/dto/UserProfileView.java:
```diff
@@ -11,7 +11,7 @@ public record UserProfileView(
         String oneLiner,
         UserRole userRole,
-        Team favoriteTeam
+        TeamView favoriteTeam
```
굿👍

csct3434 commented on src/main/java/com/example/temp/fanpool/service/impl/FindHostedFanpoolByUserServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class FindHostedFanpoolByUserServiceImpl implements FindHostedFanpoolByUserService {
+    private final UserRepository userRepository;
+    private final FanpoolRepository fanpoolRepository;
+
+    @Override
+    public FindHostedFanpoolByUserResponse doService(FindHostedFanpoolByUserCommand command) {
+        User user = userRepository.findByIdOrElseThrow(command.userId());
+        List<Fanpool> result = fanpoolRepository.findByHostUserOrderByCreatedAtDesc(user, command.pageable());
+
+        return FindHostedFanpoolByUserResponse.build(result);
+    }
+}
```
@Transactional(readonly = true) 🙏

SEONMl replied:
ㅋㅋㅋㅋㅋㅋㅋㅋ

csct3434 commented on src/main/java/com/example/temp/fanpool/service/impl/ParticipateFanpoolServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+            throw new CustomException(HttpStatus.BAD_REQUEST, "이미 참여 중입니다.");
+        }
+
+        FanpoolUser fanpoolUser = FanpoolUser.build(fanpool, user);
+        fanpool.updateCurrentNumberOfPeople();
+        fanpoolUserRepository.save(fanpoolUser);
```
동시성 처리 도전해보기 딱 좋은듯 😇

dsxl8ior replied:
잘 못들었습니다?

csct3434 commented on src/main/java/com/example/temp/user/dto/UserProfileView.java:
```diff
@@ -21,7 +22,8 @@ public UserProfileView(User user) {
                 user.getUserRole(),
-                new TeamView(user.getFavoriteTeam())
+                new TeamView(user.getFavoriteTeam()),
+                user.getHostedFanpool().size()
```
N+1 쿼리 나갈 거 같은데 countBy로 갯수만 조회해 오는 것도?

dsxl8ior replied:
하다보니까 너무 많은 정보 보내는 것 같아서 나중에 리팩하려고용

csct3434 commented on src/main/java/com/example/temp/baseball/service/impl/FindGamesByTeamServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class FindGamesByTeamServiceImpl implements FindGamesByTeamService {
+    private final TeamRepository teamRepository;
+    private final GameRepository gameRepository;
+
+    @Override
+    public FindGamesByTeamResponse doService(FindGamesByTeamCommand command) {
+        Team team = teamRepository.findByIdOrElseThrow(command.teamId());
+        List<Game> games = gameRepository.findByTeamOrderByStartDate(team);
+        return FindGamesByTeamResponse.build(games);
+    }
+}
```
@Transactional(readonly = true) 🙏

dsxl8ior replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋ ^___^;;;

csct3434 commented on src/main/java/com/example/temp/baseball/controller/BaseballGameController.java:
```diff
@@ -18,7 +18,7 @@
 @RestController
 @RequiredArgsConstructor
-@RequestMapping("/baseball-game")
+@RequestMapping("/baseball/game")
```
👍

csct3434 committed:
refactor: 레포지토리 조회 반환타입 Optional 래핑

csct3434 committed:
fix: 회원 인증 장소 레포지토리 findByUserOrElseThrow() 예외 처리 로직 오류 정정

오류: 조회 결과가 없어도 size가 0인 List가 들어오기 때문에 Optional.isEmpty()는 항상 false이다
정정: 조회 결과가 없는 경우 예외가 발생하도록 메서드의 반환 타입과 조건문을 변경

csct3434 committed:
refactor: 조회 서비스 로직에 @Transactional(reaonly=true) 어노테이션 추가

- 데이터 무결성 보장을 위한 @Transactional(reaonly=true) 어노테이션 추가

csct3434 committed:
refactor: 미사용 필드 제거
